### PR TITLE
Fix #9871 by making jl_substrtod/f whitespace-tolerant

### PR DESCRIPTION
**Pull request by IainNZ**

### CODE REVIEW

File: src/builtins.c [hunk 1]
```diff
@@ -755,7 +755,7 @@ DLLEXPORT int jl_substrtod(char *str, size_t offset, int len, double *out)
         pend = bstr+len;
     }
     *out = strtod_c(bstr, &p);
-    if ((p == bstr) || (p != pend) ||
+    if ((p == bstr) || (p != pend && !isspace((unsigned char)*p)) ||
```
Comment (IainNZ) [inline src/builtins.c]: In the case of the original bug report, `p` would be the newline character, which we allow in jl_strtod

**Comment by IainNZ** [comment]
An odd thing this fix reveals is that
```
@test float(split("0,1\nX",","))[2] == 1.0
```
passes, despite the X in there that would fail if used `float("1\nX")`. I guess I should fix that too!

**Comment by jakebolewski** [comment]
Do we want this behavior?  As a datapoint, Python allows for trailing whitespace.

```python
>>> int("1\n")
1
>>> int("1 ")
1
```

**Comment by IainNZ** [comment]
I like it, makes parsing files a lot easier (e.g. parsing a line floats from a CSV manually doesn't require a `chomp`). The current situation is broken/inconsistent, at least. Can I get an OK-to-merge from someone? This maybe should be backported if we are still doing that.

**Comment by jakebolewski** [comment]
Doesn't `jl_substrtof` have the same issue?

**Comment by IainNZ** [comment]
Good point! Pushed amended commit

File: test/strings.jl [hunk 2]
```diff
@@ -1305,3 +1305,14 @@ for T in (ASCIIString, UTF8String, UTF16String, UTF32String)
         end
     end
 end
+
+# issue #9781
+# float(SubString) wasn't tolerant of trailing whitespace, which was different
+# to "normal" strings. This also checks we aren't being too tolerant and allowing
+# any arbitrary trailing characters.
+@test float("1\n") == 1.0
```
Comment (jakebolewski) [inline test/strings.jl]: This needs to be changed to float64, but otherwise lgtm.

Comment (IainNZ) [inline test/strings.jl]: It does call float64 though on all systems, right?

Comment (jakebolewski) [inline test/strings.jl]: It does, just thought it would be good to be explicit.

Comment (IainNZ) [inline test/strings.jl]: Sure, I'll add it

**Comment by tkelman** [comment]
There was an unusual Travis failure here for OSX that looked like #9572, but someone restarted the build and now the log with the failure is gone. If a failure log is worth keeping, it might be preferable to restart the build via rebasing rather than restarting through Travis' interface.

**Comment by IainNZ** [comment]
I restarted it, sorry. Assumed it was a known issue. It did look like that though, you're right.

**Comment by pao** [comment]
Is this a backport candidate? https://github.com/JuliaLang/julia/issues/9781#issuecomment-100103581

**Comment by tkelman** [comment]
I think it should be. Needs careful conflict resolution in backporting the tests, since there are other tests nearby for recent utf8proc and reverseind which would likely fail on release-0.3.

**Comment by tkelman** [comment]
Backported in 66f081d2530aca1cfc777449f2411890d862c8ab - someone speak up before we tag 0.3.9 in a few weeks if there's any backwards compatibility concerns about this (or any other backport for that matter). Apparently nobody was paying attention to the

> This maybe should be backported if we are still doing that.

bit a few months ago, whoops.

**Pull request merged**